File: frontend/src/services/sessions.ts
import { api } from './api'
import type {
  ApiResponse,
  DeleteSessionResponse,
  PageResult,
  Session,
  SessionDetail,
  SessionSource,
} from '../types/api'

const useMock = import.meta.env.VITE_USE_MOCK === 'true'

export const sessionService = {
  async getSessions(keyword = ''): Promise<ApiResponse<PageResult<Session>>> {
    if (useMock) {
      const mockApi = await import('../mocks/mockApi')
      return mockApi.getSessions(keyword)
    }
    const response = await api.get<ApiResponse<PageResult<Session>>>('/sessions', {
      params: { keyword, page: 1, page_size: 20 },
    })
    return response.data
  },
  async createSession(source: SessionSource): Promise<ApiResponse<Session>> {
    if (useMock) {
      const mockApi = await import('../mocks/mockApi')
      return mockApi.createSession(source)
    }
    const response = await api.post<ApiResponse<Session>>('/sessions', { source })
    return response.data
  },
  async getSessionDetail(sessionId: string): Promise<ApiResponse<SessionDetail>> {
    if (useMock) {
      const mockApi = await import('../mocks/mockApi')
      return mockApi.getSessionDetail(sessionId)
    }
    const response = await api.get<ApiResponse<SessionDetail>>(`/sessions/${sessionId}`)
    return response.data
  },
  async deleteSession(sessionId: string): Promise<ApiResponse<DeleteSessionResponse>> {
    if (useMock) {
      const mockApi = await import('../mocks/mockApi')
      return mockApi.deleteSession(sessionId)
    }
    const response = await api.delete<ApiResponse<DeleteSessionResponse>>(`/sessions/${sessionId}`)
    return response.data
  },
}

File: backend/src/api/routes/sessions.py
"""Session history API routes."""

from typing import cast

from fastapi import Depends, Query
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession

from pycore.api import APIRouter
from pycore.api.responses import APIResponse, success_response

from ...api.deps import get_session
from ...models.session import (
    DeleteSessionRead,
    SessionCreate,
    SessionDetailRead,
    SessionPageRead,
    SessionRead,
)
from ...services.session_service import SessionNotFoundError, SessionService

router = APIRouter(prefix="/api/sessions", tags=["sessions"])
DB_SESSION_DEPENDENCY = Depends(get_session)


def _error(message: str, status_code: int) -> JSONResponse:
    return JSONResponse(
        status_code=status_code,
        content={"code": status_code, "message": message, "data": None},
    )


@router.get("", response_model=APIResponse[SessionPageRead])
async def list_sessions(
    keyword: str = "",
    page: int = Query(default=1),
    page_size: int = Query(default=20),
    db: AsyncSession = DB_SESSION_DEPENDENCY,
) -> APIResponse[SessionPageRead] | JSONResponse:
    """List sessions with keyword filtering."""
    try:
        data = await SessionService(db).list_sessions(keyword, page, page_size)
    except ValueError as exc:
        return _error(str(exc), 400)
    return cast(APIResponse[SessionPageRead], success_response(data=data.model_dump()))


@router.post("", response_model=APIResponse[SessionRead])
async def create_session(
    payload: SessionCreate,
    db: AsyncSession = DB_SESSION_DEPENDENCY,
) -> APIResponse[SessionRead]:
    """Create a draft session."""
    data = await SessionService(db).create_session(payload.source)
    return cast(APIResponse[SessionRead], success_response(data=data.model_dump()))


@router.get("/{session_id}", response_model=APIResponse[SessionDetailRead])
async def get_session_detail(
    session_id: str,
    db: AsyncSession = DB_SESSION_DEPENDENCY,
) -> APIResponse[SessionDetailRead] | JSONResponse:
    """Return a session detail and its messages."""
    try:
        data = await SessionService(db).get_detail(session_id)
    except SessionNotFoundError as exc:
        return _error(str(exc), 404)
    return cast(APIResponse[SessionDetailRead], success_response(data=data.model_dump()))


@router.delete("/{session_id}", response_model=APIResponse[DeleteSessionRead])
async def delete_session(
    session_id: str,
    db: AsyncSession = DB_SESSION_DEPENDENCY,
) -> APIResponse[DeleteSessionRead] | JSONResponse:
    """Delete a session."""
    try:
        data = await SessionService(db).delete_session(session_id)
    except SessionNotFoundError as exc:
        return _error(str(exc), 404)
    return cast(APIResponse[DeleteSessionRead], success_response(data=data.model_dump()))
